File: src-tauri/src/infrastructure/ffmpeg/render/wave/effects/player.rs
//! `player` — media player style: amplitude texture + progress fill + play triangle.
//! Parity: src/waves/player.ts
use crate::infrastructure::ffmpeg::render::{
    frame::{BAR_FILL, GAP_FILL, WAVE_BARS},
    pixel::{blend, fill_rect},
    wave::{WaveCtx, WaveEffect},
};

pub struct PlayerEffect;

impl WaveEffect for PlayerEffect {
    fn id(&self) -> &'static str { "player" }

    fn draw(&self, px: &mut [u8], c: &WaveCtx) {
        let n = WAVE_BARS;
        let bar_w = c.ww as f32 * BAR_FILL / n as f32;
        let gap   = c.ww as f32 * GAP_FILL / n as f32;
        let mid_y = c.wy + c.wh / 2.0;
        let progress = if c.dur > 0.0 { (c.t_sec / c.dur).min(1.0) as f32 } else { 0.0 };
        let fill_x = c.wx as f32 + c.ww as f32 * progress;

        // Background amplitude bars (dim)
        for (i, &p) in c.heights.iter().enumerate() {
            let bh = (p * c.wh * 0.7).max(3.0);
            let x0 = (c.wx as f32 + i as f32 * (bar_w + gap)) as usize;
            let x1 = (x0 as f32 + bar_w).min((c.wx + c.ww) as f32) as usize;
            let y0 = (mid_y - bh / 2.0).max(0.0) as usize;
            let y1 = (mid_y + bh / 2.0).min(c.h as f32) as usize;
            fill_rect(px, c.w, c.h, x0, y0, x1, y1, c.wc, 0.15);
        }

        // Progress-filled bars
        for (i, &p) in c.heights.iter().enumerate() {
            let bh = (p * c.wh * 0.7).max(3.0);
            let x0f = c.wx as f32 + i as f32 * (bar_w + gap);
            if x0f >= fill_x { break; }
            let x0 = x0f as usize;
            let x1 = ((x0f + bar_w).min(fill_x)).min((c.wx + c.ww) as f32) as usize;
            let y0 = (mid_y - bh / 2.0).max(0.0) as usize;
            let y1 = (mid_y + bh / 2.0).min(c.h as f32) as usize;
            let bar_alpha_step = 1.0 / bh.max(1.0);
            for y in y0..y1 {
                let local_t = (y as f32 - y0 as f32) * bar_alpha_step;
                let a = 1.0 - local_t * 0.45; // gradient: full at top, 55% at bottom
                fill_rect(px, c.w, c.h, x0, y, x1, y + 1, c.wc, a);
            }
        }

        // Scrubber line
        let lx = fill_x as usize;
        if lx + 2 <= c.wx + c.ww {
            fill_rect(px, c.w, c.h, lx, c.wy as usize, lx + 2, (c.wy + c.wh) as usize, c.wc, 0.95);
        }

        // Play triangle (left of wave zone)
        let tri_h = (c.wh * 0.38).max(6.0);
        let tri_x = c.wx.saturating_sub((tri_h * 1.2) as usize);
        for row in 0..(tri_h as usize) {
            let t = row as f32 / tri_h;
            let half_w = (t * tri_h * 0.85 / 2.0) as usize;
            let cy = (mid_y - tri_h / 2.0) as usize + row;
            let cx = tri_x + half_w.min(tri_x);
            let cx_end = (tri_x + 2 * half_w + 1).min(c.w);
            for x in cx..cx_end {
                if let Some(idx) = (cy * c.w + x).checked_mul(4) {
                    if idx + 3 < px.len() {
                        blend(px, idx, c.wc, 0.80);
                    }
                }
            }
        }
    }
}

File: apps/desktop/src/waves/player.ts
// `player` — media player style: amplitude texture + progress fill + play triangle + time text
import type { WaveEffect } from './types'
import { WAVE_BARS } from './support'

export const playerEffect: WaveEffect = {
  id: 'player',
  label: 'Player',
  desc: 'Media player',
  draw({ ctx, color, peaks, heights, waveTime, waveDur, wx, wy, ww, wh }) {
    const N = WAVE_BARS
    const bw = ww * 0.64 / N
    const gap = ww * 0.36 / N
    const midY = wy + wh / 2

    // progress 0–1
    const progress = waveDur > 0 ? Math.min(1, waveTime / waveDur) : 0

    // Background amplitude bars (dim)
    heights.forEach((p, i) => {
      const bH = Math.max(3, p * wh * 0.7)
      const x0 = wx + i * (bw + gap)
      const barW = Math.min(wx + ww - x0, bw)
      const y0 = midY - bH / 2
      ctx.fillStyle = color + '28'
      ctx.beginPath(); ctx.roundRect(x0, y0, barW, bH, 2); ctx.fill()
    })

    // Progress fill: draw bars with full color up to progress point
    const fillX = wx + ww * progress
    ctx.save()
    ctx.beginPath(); ctx.rect(wx, wy, fillX - wx, wh); ctx.clip()
    heights.forEach((p, i) => {
      const bH = Math.max(3, p * wh * 0.7)
      const x0 = wx + i * (bw + gap)
      const barW = Math.min(wx + ww - x0, bw)
      const y0 = midY - bH / 2
      const g = ctx.createLinearGradient(0, y0, 0, y0 + bH)
      g.addColorStop(0, color + 'FF'); g.addColorStop(1, color + '88')
      ctx.fillStyle = g
      ctx.beginPath(); ctx.roundRect(x0, y0, barW, bH, 2); ctx.fill()
    })
    ctx.restore()

    // Thin progress scrubber line
    ctx.fillStyle = color
    ctx.fillRect(fillX - 1, wy, 2, wh)

    // Play triangle at left edge
    const triH = wh * 0.38
    const triX = wx - triH * 1.2
    ctx.fillStyle = color + 'CC'
    ctx.beginPath()
    ctx.moveTo(triX, midY - triH / 2)
    ctx.lineTo(triX + triH * 0.85, midY)
    ctx.lineTo(triX, midY + triH / 2)
    ctx.closePath()
    ctx.fill()

    // Time text: elapsed / total (right edge)
    const fmtTime = (s: number) => {
      const m = Math.floor(s / 60)
      const sec = Math.floor(s % 60)
      return `${m}:${String(sec).padStart(2, '0')}`
    }
    const elapsed = waveDur * progress
    const timeStr = `${fmtTime(elapsed)} / ${fmtTime(waveDur)}`
    const fs = Math.max(10, wh * 0.35)
    ctx.font = `600 ${fs}px monospace`
    ctx.fillStyle = color + 'CC'
    ctx.textAlign = 'right'
    ctx.fillText(timeStr, wx + ww, midY + fs * 0.35)

    void peaks // unused but part of API
    void N
  },
}
